### PR TITLE
Feature/analysis pipeline

### PR DESCRIPTION
**Pull request by dsmits**
Leaves a lot to be desired, view it as a first prototype

### CODE REVIEW

File: v6_carrier_py/master.py [hunk 1]
```diff
@@ -105,12 +112,53 @@ def correlation_matrix(client: ClientContainerProtocol, data, keys=None, *args,
...
+    combined_df = _combine_all_node_data(client, data, keys, *args, **kwargs)
+
+    return combined_df.corr()
+
+
+def train_model(client: ClientContainerProtocol, data, pipeline: Pipeline, features: List[str], target: str, keys=None,
+                *args, **kwargs):
+    """
+    Retrieve data from nodes and train data analysis pipeline on it. Returns the performance of the resulting model.
+    TODO: How and where do we save our model?
```
Comment (svenvanderburg) [inline v6_carrier_py/master.py:125]: I guess you want to return the model? But I guess it is hard to serialize. Maybe store it somewhere and return a download link (might also need authentication)

File: v6_carrier_py/master.py [hunk 2]
```diff
@@ -105,12 +112,53 @@ def correlation_matrix(client: ClientContainerProtocol, data, keys=None, *args,
     If no keys are specified the datasets are joined on all columns with the same name.
     TODO: What if different datasets use different keys to mean the same thing? How do we specify this?
     """
-    results = _dispatch_tasks(client, data, *args, method='get_data', **kwargs)
+    combined_df = _combine_all_node_data(client, data, keys, *args, **kwargs)
+
+    return combined_df.corr()
+
+
+def train_model(client: ClientContainerProtocol, data, pipeline: Pipeline, features: List[str], target: str, keys=None,
```
Comment (svenvanderburg) [inline v6_carrier_py/master.py]: Would be nice to document arguments

Comment (svenvanderburg) [inline v6_carrier_py/master.py]: I would vote for something like `run_pipeline` instead of `train_model`

File: v6_carrier_py/master.py [hunk 3]
```diff
@@ -105,12 +112,53 @@ def correlation_matrix(client: ClientContainerProtocol, data, keys=None, *args,
...
+    """
+    try:
+        info(f'Training pipeline with the following steps: {pipeline.named_steps}')
+        results = _combine_all_node_data(client, data, keys, *args, **kwargs)
+
+        X = results[features].values
+        y = results[target].values
+
+        # Split data
+        X_train, X_test, y_train, y_test = train_test_split(X, y)
```
Comment (svenvanderburg) [inline v6_carrier_py/master.py]: This should at least use a fixed random state to enforce having the same split. 
And preferably you would want to have a dedicated test set, I guess that should be something that the data station itself specifies, but in practice you will probably also need information from other data stations. Not sure... Maybe you can add a comment to think about this?

File: v6_carrier_py/master.py [hunk 4]
```diff
@@ -105,12 +112,53 @@ def correlation_matrix(client: ClientContainerProtocol, data, keys=None, *args,
...
+        # TODO: Make metrics configurable
+        score = metrics.mean_absolute_error(y_test, predictions)
+
+        return score
+    except Exception as e:
+        # Log stacktrace
+        traceback.print_exc()
+
+
+def _combine_all_node_data(client, data, keys, *args, **kwargs) -> pd.DataFrame:
```
Comment (svenvanderburg) [inline v6_carrier_py/master.py]: Can you rename `keys` to something like: `record_identifiers` or `identifying_columns`. Something that says what it is.

Comment (svenvanderburg) [reply v6_carrier_py/master.py]: If you by the way want to enforce pseudonymization this will always be the encrypted identifier column

File: tests/resources/joined_data.csv [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
...227,45.66092650982294,1.0,143.53043900072763,340.23162745573615,1127.425782197429,28.79153528983749,46.299660441426155,0.989328159107446,16.120058210041233,27.21937909289353,787.6514673781227,45.66092650982294,1.0,124.40923706518558,211.91493715287925,1393.5342005261618,13.65097924583455,7.263227126571177,13.382344343759133,6.7743349897690734,0.0,65.76483484361296,13.381759719380298,1.0,131.01382488479263,177.42140296979005,456.30547875064,55.194879672299024,0.0,33.43563748079877,11.054275473630312,11.97900665642601,140.31049667178698,0.0,107.70916538658471,1.0,159.54646153846156,215.888,699.692,16.253076923076925,11.737076923076925,37.203692307692315,4.9310769230769225,18.65246153846154,218.92461538461532,0.0,4.0958461538461535,0.0,1.0,114.70625,95.87,942.9320833333333,49.33333333333334,0.0,0.0,0.0,0.0,0.0,0.0,0.0,35.22416666666667,9.375,1.0,155.86588979303022,391.3936405448434,1239.6159782416416,49.39521050769503,18.585795153016093,73.68680346718557,6.174464885901291,18.76191845038033,299.71274102246593,0.0,3.3886874226074646,63.31571731823811,23.077812665841144,1,2012-07-25,66,57,70,92,23,39,66,75,16,90,37,86,62,58,54,76,89,27,40,48,75,35,42,42,31,52,78,38,24,15,33,25,47,95,8,96
```
Comment (svenvanderburg) [inline tests/resources/joined_data.csv:3]: Not sure if you really need such a big complicated dataset to test the things that you want to test

File: tests/test_master.py [hunk 6]
```diff
@@ -108,10 +123,26 @@ def test_correlation_matrix_dont_mix_up_partly_matching_keys():
...
+def test_train_model_accepts_dataset():
+    dataset = load_dataset()
+    pipe = pipeline.make_pipeline(StandardScaler(), GaussianNB(priors=None))
+
+    client = create_base_mock_client()
+    client.get_results.return_value = [dataset]
+    result = master.train_model(client, None, pipe, FEATURES, TARGET, IDENTIFIER_KEYS)
+
+    print(result)
+    assert True
```
Comment (svenvanderburg) [inline tests/test_master.py]: `assert True` 🎉

File: v6_carrier_py/master.py (hunk 2 above)
Comment (svenvanderburg) [inline v6_carrier_py/master.py]: By allowing the user to add a Pipeline, the user can basically inject anything he wants. This is fine as long as you only return the metrics and you specify what kind of metrics. But it becomes tricky if you want to return the model. In that case you could quite easily steal data by executing a pipeline that .... well, steals data.

**Comment by svenvanderburg** [comment]
I've left some comments, but I guess the main question I have is: Where is this going? What is the idea? It feels like you are balancing between 2 'goals' that don't mix well:

1. Develop a general interface for a user to run a machine learning pipeline on the nodes. I.e. this will be some sort of algorithm container that handles all kind of analyses.
2. Setting up a toy algorithm to experiment with

If you want to go for 1 you need to:
- Check if this is really what a user wants to have. 
- Have a clear, generic interface to the user that is properly thought through (that takes quite some effort, this is I guess something that pysyft is trying to be)
- In terms of security you need to limit the options for the user, which is hard to do in practice since you also want it to be general.

I am not sure if you want to go in this direction, maybe something that would work better is a specific algorithm container that does 1 thing and is limitedly configurable by the user.

Option 2 is maybe also what you want to achieve, but it goes a bit against option 1.

I'm also not sure, so let's discuss!

File: v6_carrier_py/master.py (hunk 3 above)
Comment (svenvanderburg) [reply v6_carrier_py/master.py]: I guess a user would also want to have control over this...

File: v6_carrier_py/master.py (hunk 1 above)
Comment (dsmits) [reply v6_carrier_py/master.py:125]: I don't know if I want to return the model actually, It seems like something that is hard to understand whether it infringes on someones privacy. I'd rather keep it at the TSE and just return the performance, plus maybe an ID of the model you just ran. Then if the performance is satisfactory you can run another algorithm to use this model on other data.

I'm not 100% convinced about this though, I think I need to read up some more about federated learning.

Comment (dsmits) [reply v6_carrier_py/master.py:125]: As an example of a mode infringing on someones privacy. There could be an overfitted decision tree where one of the leafs only fits one person in the training set. Looking a the model could tell you a lot about that individual.

Comment (dsmits) [reply v6_carrier_py/master.py:125]: We could also return an encrypted version of the model, where the user can give it as input to an algorithm but can't decrypt it themselves.

File: v6_carrier_py/master.py (hunk 3 above)
Comment (dsmits) [reply v6_carrier_py/master.py]: I will add a comment to think about this. I don't think the data station should necessarily already have split the data in train and test, the data is more generic than that.

I agree that the user would like to have control over this, but this can come at a later stage.

Comment (dsmits) [reply v6_carrier_py/master.py]: I will add a random seed though

File: tests/resources/joined_data.csv (hunk 5 above)
Comment (dsmits) [reply tests/resources/joined_data.csv:3]: This is a part of the dataset that the master algorithm receives when running on `vantage6-local-setup`. I had a lot of issues with getting the pipeline to work on this while the iris dataset ran like a dream. But the dataset is a little large I admit, I'll remove most of the lines.

File: tests/test_master.py (hunk 6 above)
Comment (dsmits) [reply tests/test_master.py]: Are you mocking me ;P  I'm interested in whether the function succeeds at all. This test is very important for debugging. I could also check the result of the model but I'd say that value (mean absolute error at the moment) is too specific and meaningless.

File: v6_carrier_py/master.py (hunk 2 above)
Comment (dsmits) [reply v6_carrier_py/master.py]: Yeah I'm aware of that. This is basically what Chang had implemented. I see it as a starting. With pythons duck typing you can basically package any type of malicious code as a "data analysis pipeline".

Still, I think it is good if this function is generic for now. Maybe later it will be wrapped by another algorithm that constructs a pipeline server-side based on user input.

**Review by svenvanderburg — APPROVED** [review]
Looks good! Nice discussions :)
